account_force_early_discount/models/account_move.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2019-2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    display_force_early_discount = fields.Boolean(
+        compute="_compute_display_force_early_discount"
+    )
+    force_early_discount = fields.Boolean(
+        "Force early payment discount",
+        default=False,
+        help="If marked, early payment discount will be applied even if the "
+        "discount date is passed",
+    )
+
+    def _get_display_force_early_discount(self):
+        self.ensure_one()
+        if self.state == "draft":
+            if (
+                self.invoice_payment_term_id.discount_days
+                and self.invoice_payment_term_id.discount_percentage
+            ):
+                return True
+        elif self.state == "posted":
+            first_payment_term_line = self._get_first_payment_term_line()
+            if (
+                first_payment_term_line.discount_date
+                and first_payment_term_line.discount_amount_currency
+            ):
+                return True
+        return False
+
+    @api.depends(
+        "invoice_payment_term_id",
+        "invoice_payment_term_id.early_discount",
+        "force_early_discount",
+        "state",
+    )
+    def _compute_display_force_early_discount(self):
+        """Compute discount financial discount fields"""
+        for rec in self:
+            rec.display_force_early_discount = rec._get_display_force_early_discount()
+
+    def _get_first_payment_term_line(self):
+        self.ensure_one()
+        payment_term_lines = self.line_ids.filtered(
+            lambda line: line.display_type == "payment_term"
+        )
+        return fields.first(payment_term_lines.sorted("date_maturity"))
```

Comment:
For a long time I also used `fields.first`.
But then I realized, by reading its docstring, that it really doesn't make sense to use it in most cases.

```
def first(records):
    """ Return the first record in ``records``, with the same prefetching. """
    return next(iter(records)) if len(records) > 1 else records
```

Essentially, you get the first record with the prefetching set to all records.. but you usually don't really care about the other records so why would you want them prefetched?

Instead, the pythonic way works well and better fits most usecases:

```suggestion
        return payment_term_lines.sorted("date_maturity")[:1]
```

That'll give you just the first record, with no prefetching on the others

Reply:
AFAIR we used to recommend using `fields.first` to avoid an error when slicing using `[0]`, but I'm not sure why we never came up with this before :confused: